### DISCUSSION TITLE
Requesting for WT_Perf

### DISCUSSION THREAD

**Discussion by KarimMohammadd** [OpenFAST/openfast#1162] [General]
Hello everyone
We know that WT_Perf is a program that uses blade-element momentum (BEM) theory to predict the performance of wind turbines. Does anyone know where I can get WT_Perf v3.05?

**Comment by jjonkman** [comment]
WT_Perf is no longer supported by NREL, but has been effectively replaced by the standalone driver for AeroDyn v15 within OpenFAST or CCBlade within WISDEM.

Best regards,

**Reply by LeBete02** [reply]
Hello, I am fairly new to using openfast. I'm curious how I would generate Cp values for an MPC controller.  I have attached below the controllers that I am currently using (which is the same as the Bladed .dll file).

 function [PitchControl, IntSpdErr]= pitchController(GenSpeed, LastIntSpdErr)

% Define constants and initial values
% OnePlusEps = 1.00000001;  % Slightly greater than unity
PC_DT = 0.00625;  % Communication interval for pitch controller, sec
PC_KI = 0.008068634;  % Integral gain for pitch controller at rated pitch (zero), (-)
PC_KP = 0.01882681;  % Proportional gain for pitch controller at rated pitch (zero), sec
PC_MaxPit = 1.570796;  % Maximum pitch setting in pitch controller, rad
PC_MinPit = 1 * pi/180;  % Minimum pitch setting in pitch controller, rad
PC_RefSpd = 122.9096;  % Desired (reference) HSS speed for pitch controller, rad/s
NumBl = 3;  % Number of blades

ElapTime = PC_DT;  % Elapsed time since the last call to the controller
GK = 1.0;  % Gain scheduling correction factor
IntSpdErr = LastIntSpdErr; % Current integral of speed error w.r.t. time
PitCom = zeros(NumBl, 1);  % Overall command pitch for each blade, rad
GenSpeedF = GenSpeed;

% Compute the current speed error and its integral w.r.t. time; saturate the integral term using the pitch angle limits
SpdErr = GenSpeedF - PC_RefSpd;
IntSpdErr = IntSpdErr + SpdErr * ElapTime;
IntSpdErr = min(max(IntSpdErr, PC_MinPit / (GK * PC_KI)), PC_MaxPit / (GK * PC_KI));

% Compute the pitch commands associated with the proportional and integral gains
PitComP = GK * PC_KP * SpdErr;
PitComI = GK * PC_KI * IntSpdErr;

% Superimpose the individual commands to get the total pitch command; saturate the overall command using the pitch angle limits
PitComT = PitComP + PitComI;
PitComT = min(max(PitComT, PC_MinPit), PC_MaxPit);

% Saturate the overall commanded pitch using the pitch rate limit
for K = 1:NumBl
    PitCom(K) =  PitComT;
end
        
PitchControl = PitCom(1);
        
end

function [GenTrq, GenPwr] = generatorTorqueController(GenSpeedF, PitCom, LastGenTrq)
    %% Constants and Parameters
    VS_RtPwr = 5e6;                      % Rated generator power in Region 3, Watts
    VS_RtGnSp = 121.6805;                % Rated generator speed (HSS side), rad/s
    VS_Rgn3MP = 0.01745329;              % Minimum pitch angle for Region 3, rad
    VS_CtInSp = 70.16224;                % Transitional generator speed for Region 1 and 1 1/2, rad/s
    VS_Rgn2Sp = 91.21091;                % Transitional generator speed for Region 1 1/2 and 2, rad/s
    VS_Rgn2K = 2.332287;                 % Generator torque constant in Region 2, N-m/(rad/s)^2
    VS_MaxTq = 43093.55;                 % Maximum generator torque in Region 3, N-m
    VS_MaxRat = 15000.0;                 % Maximum torque rate (in absolute value) in torque controller, N-m/s
    VS_SlPc = 10.0;                      % Rated generator slip percentage in Region 2 1/2, %
    ElapTime = 0.00625;                  % Constant elapsed time, s
    GenPwr = 5e6;                        % Rated Generator Power, W

    %% Initialize LastGenTrq on the first pass only
%     if isempty(LastGenTrq)
%         LastGenTrq = 0.0;
%     end
    
    %% Calculate Further Parameters 
    
    % Calculate synchronous speed (VS_SySp)
    VS_SySp = VS_RtGnSp / (1.0 + 0.01 * VS_SlPc);                           % Synchronous speed of region 2 1/2 induction generator, rad/s.
    
    % Calculate torque/speed slope of region 2 1/2 (VS_Slope25)
    VS_Slope25 = (VS_RtPwr / VS_RtGnSp) / (VS_RtGnSp - VS_SySp);            % Torque/speed slope of region 2 1/2 induction generator, N-m/(rad/s).

    % Calculate transitional generator speed (VS_TrGnSp)
    if GenSpeedF >= VS_Rgn2Sp
        VS_TrGnSp = (VS_Slope25 - sqrt(VS_Slope25 * ...
            (VS_Slope25 - 4.0 * VS_Rgn2K * VS_SySp))) / (2.0 * VS_Rgn2K);
    else
        VS_TrGnSp = VS_SySp;                                                % Transitional generator speed (HSS side) between regions 2 and 2 1/2, rad/s.
    end
    
    
    
    %% Main Control Body

    % Compute the generator torque based on wind turbine region
    if (GenSpeedF >= VS_RtGnSp) || (PitCom >= VS_Rgn3MP)
        % Region 3: Power is constant
        GenTrq = VS_RtPwr / GenSpeedF;
    elseif GenSpeedF <= VS_CtInSp
        % Region 1: Torque is zero
        GenTrq = 0.0;
    elseif GenSpeedF < VS_Rgn2Sp
        % Region 1 1/2: Linear ramp in torque from zero to optimal
        GenTrq = (GenSpeedF - VS_CtInSp) * ((VS_Rgn2K * GenSpeedF * GenSpeedF) / (VS_Rgn2Sp - VS_CtInSp));
    elseif GenSpeedF < VS_TrGnSp
        % Region 2: Optimal torque is proportional to the square of the generator speed
        GenTrq = VS_Rgn2K * GenSpeedF * GenSpeedF;
    else
        % Region 2 1/2: Simple induction generator transition region
        GenTrq = VS_Slope25 * (GenSpeedF - VS_SySp);
    end

    % Saturate the commanded torque using the maximum torque limit
    GenTrq = min(GenTrq, VS_MaxTq);

    % Calculate torque rate (unsaturated)
    TrqRate = (GenTrq - LastGenTrq) / ElapTime;

    % Saturate the torque rate using its maximum absolute value
    TrqRate = min(max(TrqRate, -VS_MaxRat), VS_MaxRat);

    % Saturate the command using the torque rate limit
    GenTrq = LastGenTrq + TrqRate * ElapTime;
end

![image](https://github.com/OpenFAST/openfast/assets/96494235/0f1496a7-c20d-4439-a41a-3b855d8aaf3c)

**Reply by jjonkman** [reply]
Dear @LeBete02,

I'm not sure I really understand your question.  Are you asking how to obtain a a surface of power coefficient as a function of tip-speed ratio and blade-pitch angle?

Best regards,

**Comment by icnosvbovwb** [comment]
Dear @jjonkman
I want  to use the WT_Perf v3.05 to create performance_ccblade，Can you provide the WT_Perf v3.05？

**Reply by jjonkman** [reply]
Dear @icnosvbovwb,

What do you mean when you say, "create performance_ccblade."

Is there a reason you can't use the standalone driver for AeroDyn v15 within OpenFAST or CCBlade within WISDEM as I recommended above?

Best regards,

**Comment by icnosvbovwb** [comment]
Dear @jjonkman 
![image](https://user-images.githubusercontent.com/107483730/204093841-9bafb1bf-0334-4a55-9a60-8541529c6070.png)
Thank you for reading my question in your busy schedule. Forgive me for not expressing my meaning clearly. I want to get the performance of 5MW_ CCblade, but I'm not sure how to get it. I saw the method of using CCBlade within WISDEM in the forum, but I need to use python. I'm not familiar with python. Before that, I saw that WT_ Perf v3.05 was used to generate ccblade  in the forum, so what quick method do you have to get 5MW performance_ ccblade？

**Reply by jjonkman** [reply]
Dear @icnosvbovwb,

It appears that the file you refer to was generated with CCBlade, which I agree requires python.  If you prefer to not use python, you can use the standalone AeroDyn driver to generate similar aerodynamic performance data.

Best regards,

**Comment by icnosvbovwb** [comment]
Dear @jjonkman 
Can you give me the standalone driver for AeroDyn v15? Can it generate the following data corresponding to 3-25m/s wind speed?
Wind speed (m/s)  Rotor rotational speed (rpm) 	 Pitch angle (deg) 	 Rotor electrical power (W) 	 Rotor aerodynamic power (W) 	 Rotor thrust (N) 	 Rotor aerodynamic torque (Nm) 	 Blade root flapwise moment (Nm)	 Rotor electrical power coefficient (-) 	 Rotor aerodynamic power coefficient (-) 	 Rotor thrust coefficient (-) 	 Rotor aerodynamic torque coefficient (-)

**Reply by jjonkman** [reply]
Dear @icnosvbovwb,

NREL does not provide precompiled binary executables for the standalone drivers of OpenFAST modules, but the source code and compile scripts are all provided in the repository (https://github.com/OpenFAST/openfast) so that you can compile yourself.  If you are Windows user, I find Intel Fortran with Visual Studio easiest to use for compiling (the hardest part is installing the compiler).

Yes, the standalone AeroDyn driver can output everything you asked about (these are all outputs from AeroDyn), except the electrical power (there is no generator model in standalone AeroDyn) and the blade root flapwise moment (there is no structural mass in AeroDyn).

Best regards,

**Comment by icnosvbovwb** [comment]
Dear @jjonkman 
Or can I use openfast to calculate all the parameters of the FAST Certification Test #18: NREL 5.0 MW Baseline Wind Turbine (Onshore) ? But the only change is the steady wind, such as setting the InflowWind to 3m/s to 25m/s and the time lasts for 100s to ensure all the parameters value stability, but many operations are required, such as 3.1m/s, 3.2m/s, 3.3m/s....to 25m/s. I'm not sure if the idea right.

**Reply by jjonkman** [reply]
Dear @icnosvbovwb,

Yes, you can certainly do that, but it will take more effort and computational time than running the standalone AeroDyn driver.

Best regards,

**Comment by icnosvbovwb** [comment]
Dear @jjonkman
![image](https://user-images.githubusercontent.com/107483730/204139194-66e6bfef-2deb-4e66-8b76-748293f016a0.png)
Can all the parameters get the three tables shown in your paper generate by the standalone AeroDyn driver? I'm not sure if it can simulate 3, 4, 5 to 25m/s by using the standalone AeroDyn driver. If it can't, I can test all parameters by changing 3.1, 3.2, 3.3m/s to 25m/s working conditions.

**Reply by jjonkman** [reply]
Dear @icnosvbovwb,

The standalone AeroDyn driver assumes a rigid wind turbine with no active control, so, you can't get the related outputs such as the blade and tower deflections and electrical power output from the standalone AeroDyn driver.  (You can get these using AeroDyn coupled to OpenFAST.)

The standalone AeroDyn driver can calculate the aerodynamic performance (e.g., Cp, Ct, Cq surface as a function of TSR and pitch) in the absence of structural deflections or active control.

Best regards,

**Comment by icnosvbovwb** [comment]
Dear @jjonkman 
Thanks for your reply. So I can test all parameters by changing 3, 4, 5m/s to 25m/s working conditions, which can obtain the parameter data corresponding to each wind speed (3m/s, 4m/s, 5m/s to 25m/s). I can connect the data corresponding to each wind speed separately, so I guess I can get the graphs of the three tables in your paper.

**Comment by mohsamaz** [comment]
Hello everyone,
I want to generate Cp,Ct,Cq surface for my specific turbine blade.
i am using WISDEM,(example 10_ccblade).

I edited yaml and code respect to my bade and I run it , I got Cp-Lambda curve plot and distributed aerodynamicloads-blade fraction and also I got this message:

NaNs at 55/57: 1e-06 0.0 1.5707963267948966
NaNs at 56/57: 1e-06 0.0 1.5707963267948966

1-what is the problem or error? 

2-also I want to know how can I get Cp,Ct,Cq surface from this (example 10_ccblade)?
should I run it in range of TSR [2,20] and pitch [0,25]

lam_range  = (2.0, 14.0)  # TSR range for CP surface
beta_range = (-4.0, 8.0)  # pitch range [deg] for CP surface

3-In airfoil section of yaml file, coordinates (x,y) is also required.
why is it needed?

thank you in advance for your help.
Best regards,

**Reply by jjonkman** [reply]
This sounds like question better posted in the CCBlade repository: https://github.com/WISDEM/CCBlade/issues.
